pkg/virt-handler/cgroup/util.go

```diff
@@ -84,6 +91,66 @@ func addCurrentRules(currentRules, newRules []*devices.Rule) ([]*devices.Rule, e
 	return newRules, nil
 }
 
+func generateDeviceRulesForVMI(vmi *v1.VirtualMachineInstance, isolationRes isolation.IsolationResult) ([]*devices.Rule, error) {
+	mountRoot, err := isolationRes.MountRoot()
+	if err != nil {
+		return nil, err
+	}
+
+	var vmiDeviceRules []*devices.Rule
+	for _, volume := range vmi.Spec.Volumes {
+		switch {
+		case volume.VolumeSource.PersistentVolumeClaim != nil:
+			if volume.VolumeSource.PersistentVolumeClaim.Hotpluggable {
+				continue
```

Comment:
is it have problem in below situation:
1. hotplug volume1
2. virt-handler restart
3. hotplug volume2

in step 3 will not add rule for volume1?

Reply:
This code only checks the volumes specified explicitly in the VMI spec. Hotplug volumes are handled separately. You can find it in the commit 58f75453166ad5107dfb6eeb12ef283873e14e65 (`Hotplug volume mounter: set cgroup manager as arg`). Also, the test added in 946fe82c4dc1e7c3de85f0470c06443fcae33c90 (`tests: Ensure access to block devices is preserved`) runs this scenario and verifies that it works as expected.